### PR TITLE
Use Data.Set.alterF

### PR DESCRIPTION
**Pull request by treeowl**

### CODE REVIEW

**Comment by treeowl** [comment]
@RyanGlScott, do you want me to bump the lower bound on `containers` instead of using CPP?

**Comment by phadej** [comment]
>  do you want me to bump the lower bound on containers instead of using CPP?

We shouldn't. It is less pain to everyone if we support `containers` versions bundled with GHCs we support.

**Comment by phadej** [comment]
Is there `IntSet.alterF`?

**Comment by treeowl** [comment]
I don't remember. I'll take a look.

**Comment by treeowl** [comment]
@phadej Yes, there is an `alterF` for `IntSet`, but it's based entirely on API functions:

```haskell
alterF :: Functor f => (Bool -> f Bool) -> Key -> IntSet -> f IntSet
alterF f k s = fmap choose (f member_)
  where
    member_ = member k s

    (inserted, deleted)
      | member_   = (s         , delete k s)
      | otherwise = (insert k s, s         )

    choose True  = inserted
    choose False = deleted
```

So I guess it's a matter of taste whether to use it when available or not. We probably *shouldn't* use it for `at`, since that leads to an extra `fmap isJust`.

**Comment by phadej** [comment]
I'd be in favor of having similar implementations for `IntSet` and `Set` **or** a proper note explaining the difference (and perfectly illustrating it with non contrived example).

I don't think that extra `fmap` is noticeable for e.g. `set` implementation.

**Comment by treeowl** [comment]
`Set` and `IntSet` are quite different in this regard. Going into a `Set` two independent times means performing (possibly expensive) comparisons along the way twice, so we avoid that. For `IntSet`, all the branches are cheap. There are all sorts of trade-offs; `containers` aims to be not-awful in the case of expensive `compare` and expensive `fmap`, while also trying to be pretty fast in easier cases.

**Comment by phadej** [comment]
Expensive `fmap` + `at @(Set _)` is a speculation. (If someone cares, they should use `contains`). Please make implementation uniform. Or provide method with `at` signature in `containers`.

I won't remember this discussion few months from now. If you won't change it, the future me will.

**Comment by treeowl** [comment]
I do not understand your last remark. I'm happy to comment the source in detail on the reasoning behind the implementation.

**Comment by phadej** [comment]
> I do not understand your last remark. I'm happy to comment the source in detail on the reasoning behind the implementation.

That is fine too.

File: src/Control/Lens/At.hs [hunk 1]
```diff
@@ -473,17 +496,32 @@ instance (Eq k, Hashable k) => At (HashMap k a) where
...
+  -- Unlike a `Set`, we don't have to worry about expensive
+  -- comparisons from descending multiple times into an
+  -- `IntSet`. We are careful to share the results of insertion
+  -- or deletion across multiple positions in the `Functor`.
+  at k f s = fmap choose (f (guard member_))
+    where
+      member_ = IntSet.member k s
+
+      (inserted, deleted)
+        | member_   = (s         , IntSet.delete k s)
```
Comment (phadej) [inline src/Control/Lens/At.hs:509]: aligning is off, probably better to eat whitespace.

File: src/Control/Lens/At.hs [hunk 2]
```diff
@@ -473,17 +496,32 @@ instance (Eq k, Hashable k) => At (HashMap k a) where
   {-# INLINE at #-}
 
 instance At IntSet where
-  at k f m = f mv <&> \r -> case r of
-    Nothing -> maybe m (const (IntSet.delete k m)) mv
-    Just () -> IntSet.insert k m
-    where mv = if IntSet.member k m then Just () else Nothing
+  -- This is a copy of the implementation of `IntSet.alterF`.
```
Comment (phadej) [inline src/Control/Lens/At.hs:499]: *modified* copy?

**Comment by phadej** [comment]
Looks good. Do you mind adding the changelog entry, then this can be merged.

**Comment by treeowl** [comment]
@phadej, I removed useless reinsertion in the implementation of `ix` for `IntSet`. For `Set` and `HashSet`, things are a bit more subtle; should `ix` be able to replace a key with a different one that compares `==` to it? That's what it currently does.

**Comment by phadej** [comment]
Re `ix`, it's behavior should be `ix i = at i . _Just`. That doesn't re-insert key, does it?

**Comment by treeowl** [comment]
I don't believe so.

**Comment by phadej** [comment]
Then `ix` should reinsert either (And I think it's better that way too).

**Comment by treeowl** [comment]
@phadej, all good now?

**Comment by treeowl** [comment]
Thanks. Now I should probably add an `adjustA` to `containers` and `unordered-containers` to deal with map `Ix` better. Bleh.

**Pull request merged**